app/auth/authV2/actions/RetrieveClientDataFromCookies.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package auth.authV2.actions
+
+import config.{AgentItvcErrorHandler, FrontendAppConfig}
+import controllers.agent.routes
+import controllers.agent.sessionUtils.SessionKeys
+import models.sessionData.SessionDataGetResponse.SessionDataNotFound
+import play.api.Logger
+import play.api.mvc.Results.Redirect
+import play.api.mvc.{ActionRefiner, Request, Result}
+import services.SessionDataService
+import services.agent.ClientDetailsService
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class RetrieveClientDataFromCookies @Inject()(sessionDataService: SessionDataService,
+                                              errorHandler: AgentItvcErrorHandler)
+                                             (implicit val executionContext: ExecutionContext) extends ActionRefiner[Request, ClientDataRequest] {
+
+  lazy val logger: Logger = Logger(getClass)
+
+  lazy val noClientDetailsRoute: Result = Redirect(routes.EnterClientsUTRController.show)
+
+  override protected def refine[A](request: Request[A]): Future[Either[Result, ClientDataRequest[A]]] = {
+
+    implicit val r: Request[A] = request
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter
+      .fromRequestAndSession(request, request.session)
+
+    sessionDataService.getSessionData(useCookie = true).map {
+      case Right(sessionData) => Right(ClientDataRequest(
+        sessionData.mtditid,
+        r.session.get(SessionKeys.clientFirstName),
+        r.session.get(SessionKeys.clientLastName),
```

Comment:
Should you still be using the session cookie for storing the clientName? Every other request (requiring an enrolled user) calls Citizen details to get the client Name and also looking at the confluence page linked in your ticket it has clientName as something that shouldn't be stored in the cookie.

Reply:
the only place where we use this predicate is within asMTDAgentWithUnconfirmedClient / ConfirmClientUTRController;
at this stage we don't know yet client details as client not yet selected, so old logic is trying to read details from existing browser cookies, which will not be available after migration/but at this stage we just fallback to use cookies and we not yet able to call service at this stage;

Reply:
We'll have to change to using Citizen Details eventually, so I'm happy to switch over to using it now if you're ok with that @avelx

Reply:
I've now changed RetrieveClientDetails so we don't need RetrieveClientDetailsFromCookies anymore, so both methods will now use the citizen details service